modules/internal/ArrayViewSlice.chpl

```diff
@@ -0,0 +1,391 @@
+/*
+ * Copyright 2004-2017 Cray Inc.
+ * Other additional copyright holders may be indicated within.
+ *
+ * The entirety of this work is licensed under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ *
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+// ArrayViewSlice
+//
+// This module provides an array domain map class that is used to
+// represent slices of another array via a domain.
+//
+module ArrayViewSlice {
+
+  //
+  // The class representing a slice of an array.  Like other array
+  // class implementations, it supports the standard dsi interface.
+  //
+  class ArrayViewSliceArr: BaseArr {
+    type eltType;  // see note on commented-out proc eltType below...
+
+    // the representation of the slicing domain
+    //
+    // TODO: Can we privatize upon creation of the array-view slice and cache
+    // the results?
+    const _DomPid;
+    const dom; // Seems like the compiler requires a field called 'dom'...
+
+    // the representation of the sliced array
+    const _ArrPid;
+    const _ArrInstance;
+
+    // If this is an array view on a DefaultRectangular array
+    // (eventually...), the indexCache provides a mean of directly
+    // accessing the array's ddata to avoid indirection overheads
+    // through the array field above.
+    const indexCache = buildIndexCache();
+
+
+    //
+    // standard generic aspects of arrays
+    //
+
+    // these could be fields, but indirecting works just as well and
+    // makes the class less generic.
+    proc idxType type return dom.idxType;
+    proc rank param return arr.rank;
+
+    // The following seems like it ought to work, but it causes an
+    // error in the compiler for non-devel mode...  presumably due to
+    // a direct query of eltType in the compiler(?).  As a TODO we
+    // might want to hunt this down in the future...
+    //
+    //  proc eltType type return arr.eltType;
+
+
+    //
+    // introspection routine used elsewhere to filter array views
+    //
+    // TODO: Could this be replaced with more type-based introspection?
+    // I shied away from it since this is a generic class, but there
+    // must be (or should be) some way to do it without relying on
+    // methods like this...
+    //
```

Comment:
The downside of the above, to me, is creation of more global symbols.  Can you articulate what makes the above better for you?  (my guess: avoids relying on statically evaluated method overrides that won't actually dynamically dispatch, so can only be used in a generic context?).

I like the cast approach better, if that works (see my comment above for where my concern that it wouldn't came from -- though I didn't experiment with it to confirm my fears...  haven't had time yet).